File: 1004.go
package main

func longestOnes(nums []int, k int) int {
	l := 0
	r := 0

	zero_count := 0
	maxLength := 0

	for r < len(nums) {
		if nums[r] == 0 {
			zero_count++
		}
		r++

		for l < len(nums) && zero_count > k {
			if nums[l] == 0 {
				zero_count--
			}
			l++
		}

		maxLength = max(maxLength, r-l)
	}

	return maxLength
}

File: 1004.c
#include "shdstd.h"
#define MAX(x, y) (((x) > (y)) ? (x) : (y))
int longestOnes(int *nums, int size, int k)
{
	int zero_count = 0;
	int l = 0;
	int r = 0;
	int max_length;

	while (r < size) {
		zero_count += nums[r++] == 0;

		while (l < size && zero_count > k) {
			zero_count -= nums[l++] == 0;
		}

		max_length = MAX(max_length, r - l);
	}

	return max_length;
}
